### PR TITLE
Prevent from future leaks of goroutines in integration tests

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #108483

```release-note
NONE
```

/kind cleanup
/sig testing
/priority important-soon

### CODE REVIEW

File: test/integration/framework/etcd.go [hunk 1]
```diff
@@ -192,23 +193,27 @@ func EtcdMain(tests func() int) {
...
-		time.Sleep(100 * time.Millisecond)
+		return false, nil
+	}
+
+	// It generally takes visibly less than 1s to finish all goroutines.
+	// But we keep the limit higher to account for cpu-starved environments.
+	if err := wait.Poll(100*time.Millisecond, 5*time.Second, checkNumberOfGoroutines); err != nil {
+		after := runtime.NumGoroutine()
+		klog.Fatalf("unexpected number of goroutines: before: %d after %d", before, after)
+		os.Exit(result)
```
Comment (aojea) [inline test/integration/framework/etcd.go:215]: if we use fatal I think that we never reach os.Exit

Comment (aojea) [reply test/integration/framework/etcd.go:215]: the os.Exit(result) should go out of the err!=nil

Comment (wojtek-t) [reply test/integration/framework/etcd.go]: Sure - put that in the wrong place. Fixed.

**Comment by wojtek-t** [comment]
@aojea - PTAL

File: test/integration/framework/etcd.go [hunk 2]
```diff
@@ -192,22 +193,25 @@ func EtcdMain(tests func() int) {
...
 		runtime.Gosched()
-		time.Sleep(100 * time.Millisecond)
+		return false, nil
+	}
+
+	// It generally takes visibly less than 1s to finish all goroutines.
+	// But we keep the limit higher to account for cpu-starved environments.
+	if err := wait.Poll(100*time.Millisecond, 5*time.Second, checkNumberOfGoroutines); err != nil {
+		after := runtime.NumGoroutine()
+		klog.Fatalf("unexpected number of goroutines: before: %d after %d", before, after)
```
Comment (aojea) [inline test/integration/framework/etcd.go:213]: > // Fatal logs to the FATAL, ERROR, WARNING, and INFO logs,
> // prints stack trace(s), then calls OsExit(255).

I think that this is safe

**Comment by aojea** [comment]
/lgtm
/retest

let's go

**Pull request merged**